### DISCUSSION TITLE
Decouple the model from the implementation?

### DISCUSSION THREAD

**Discussion by pvcnt** [perses/perses#2474] [Ideas]
Hi there, at my current job, I feel the need to have a better dashboarding model than what Grafana is offering. I've been following this project for a couple of months already, and it seems to fit the need of having a user-friendly model (as opposed to Grafana's largely undocumented one, and difficult to process). However, I'm currently not that much interested in the UI or CLI, but mostly in the model (instead of reinventing my own), and would like to be able to use Python to manipulate it.

My question is about whether it would make sense for the project to separate the model specification from its implementation? The specification could become some sort of standard, but different implementations of the server/UI/CLI could exist. In particular, I was hoping to find some interoperable schema in order to easily generate an SDK in Python. I would love it if the schemas were available in a repository, periodically updated, in various formats including OpenAPI (which is more standard than Cue, which is really only usable in Go), allowing for developers to generate SDKs, build their own implementations, etc.

**Comment by Nexucis** [comment]
Hello there,

So today we are generating the cuelang files based on the Golang implementations. I would say it wouldn't hurt to provide the specification if we are able to generate based on the Golang implementations as it is the main source.

Maybe it's not that hard to provide/generate a JSON schema.

Questions from my side:

- When you are talking about specifications, are you talking about the dashboard definition or are you talking about any resources definition used by Perses ? Could be datasource, dashboard, project ...etc.
- The dashboard definition is partial as we are plugin oriented. Are you expected that the specifications we could provide contains also the spec of every plugin we are supporting ?

**Reply by pvcnt** [reply]
To answer your questions:
* I am mostly interested in the dashboard specification, but I suppose that if you were able to publish it, publish the specification of other resources/APIs would not be much more complex.
* That is a good point... I suppose that you could make available the specification with all "official" plugins (Prometheus, Tempo, etc.) in a first step, but I'm not sure how to allow other plugins in that case, since we would end up with non-extensible unions of types. I would actually need a custom datasource for example in my case (we do not use Prometheus)...

**Comment by AntoineThebaud** [comment]
Side note about CUE: 

> One of CUE’s goals is to act as an interlingua: a bidirectional bridge between all the formats that CUE speaks, linking constraints and data sources of truth, no matter where they exist.

-> Being able to translate/encode CUE into more standard schema formats like JSON Schema and OpenAPI is in the pipe. See
- [How CUE works with JSON Schema](https://cuelang.org/docs/concept/how-cue-works-with-json-schema/#future-plans)
- [How CUE works with OpenAPI](https://cuelang.org/docs/concept/how-cue-works-with-openapi/#future-plans)

**Reply by Nexucis** [reply]
Interesting @AntoineThebaud. Thank you for the info !

@pvcnt would you like to help us to generate the dashboard definition in a json schema ?

**Reply by pvcnt** [reply]
Sorry, I didn't follow up on this. I'm not very knowledgeable in Go or Cue, but it would interest me to try doing it. It seems that exporting to OpenAPI is at least partially supported, but export to JSON Schema is not supported, am I correct?

My next question is how to best materialise this specification? For example, Grafana relies on an external repository with a dump of the schema (and other related artifacts such as the SDKs) for each release. I do not know how you handle SDK versioning, if that is handled.

**Reply by AntoineThebaud** [reply]
> It seems that exporting to OpenAPI is at least partially supported, but export to JSON Schema is not supported, am I correct?

Yes correct.

**Reply by Nexucis** [reply]
To simplify the work on this part, I think we should generate the spec of the dashboard without the various plugins.

Then regarding where it should land, I think for the moment it should be on the main repo `perses/perses` to ensure it is following the Perses version. Also because as I said, the specs are written in Go and we won't move the Go struct out of `perses/perses` for the moment.

Once this is done, and depending how it is done, we can think to generate it for each plugins on the repo `perses/plugins`. But if we do that, I guess we will need a tool that would be able to merge the various schema from the various source. Perhaps a tool that can be embedded by the server, so the server can provide the complete spec of the dashboard based on the various plugins installed.

**Reply by Nexucis** [reply]
It looks json schema generation based on cuelang is feasible with last cuelang released: https://github.com/cue-lang/cue/releases/tag/v0.12.0

**Reply by Nexucis** [reply]
schema generation is handled with this issue: https://github.com/perses/perses/issues/3596
